python/diff_frames.py

```diff
@@ -0,0 +1,160 @@
+
+Hunk = namedtuple('Hunk', ('start_a', 'end_a', 'start_b', 'end_b'))
+
+
+def duration(sndfile):
+    return sndfile.frames() / sndfile.samplerate()
+
+
+def fmt_seconds(t):
+    hours, remainder = divmod(t, 3600)
+    minutes, seconds = divmod(remainder, 60)
+    if not hours:
+        if not minutes:
+            return '{:.2f}'.format(seconds)
+        return '{:.0f}:{:.1f}'.format(minutes, seconds)
+    return '{:.0f}:{:.0f}:{.0f}'.format(hours, minutes, seconds)
+
+
+class DiffApp:
+    def __init__(self, filename_a, filename_b, loop=None):
+        self.filename_a = filename_a
+        self.filename_b = filename_b
+        self._psf_a = None
+        self._psf_b = None
+        self._diff = None
+        self._len = None
+
+        self._loop = loop or asyncio.get_event_loop()
+        self._terminal = LinePrintingTerminal()
+        self._common_fmt = self._terminal.yellow
+        self._insertion_fmt = self._terminal.green
+        self._loop.add_reader(self._terminal.infile, self._handle_input)
+        self._keyboard = Keyboard()
+        self.bindings = {}
+
+    @asyncio.coroutine
+    def __call__(self):
+        diff = yield from self._do_diff()
+        self._diff, a_offset, b_offset = self._process_diff(diff)
+        self._psf_a = pysndfile.PySndfile(self.filename_a)
+        self._psf_b = pysndfile.PySndfile(self.filename_b)
+        self._len = max(
+            self._psf_a.frames() + b_offset,
+            self._psf_b.frames() + a_offset)
+        with self._terminal.unbuffered_input(), (
+                self._terminal.nonblocking_input()), (
+                self._terminal.hidden_cursor()):
+            self._draw()
+            try:
+                while True:
+                    yield from asyncio.sleep(1)
+            finally:
+                print('\n', end='')  # Yuck :-(
+
+    def _handle_input(self):
+        key = self._keyboard[self._terminal.infile.read()]
+        action = self.bindings.get(key, lambda: None)
+        action()
+        self._draw()
+
+    @asyncio.coroutine
+    def _do_diff(self):
+        '''Calls the (fast) C-based binary diffing code to get the differences
+        between the two audio files.
+        '''
+        process = yield from asyncio.create_subprocess_exec(
+            bdiff_frames_path, self.filename_a, self.filename_b,
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = yield from process.communicate()
+        if process.returncode:
+            raise RuntimeError(stderr)
+        else:
+            return tuple(
+                Hunk(*map(int, l.split())) for l in stdout.splitlines())
+
+    @staticmethod
+    def _process_diff(diff):
+        '''Translates the raw absolute frame references in the given diff into a
+        common "virtual" space where like sections are aligned.
+        '''
+        a_offset = 0
+        b_offset = 0
+        # FIXME: is there a way to do this without an intermediate list?
```

Comment:
Depends how you slice it, I think you can make it into a generator that you can pass to tuple removing the intermediary.